### PR TITLE
Docs additions, field common methods

### PR DESCRIPTION
**Pull request by arogachev**
| Q             | A
| ------------- | ---
| Is bugfix?    | ❌
| New feature?  | ❌
| Breaks BC?    | ❌

### CODE REVIEW

File: docs/guide/en/field-common-methods/base-field/container.md [hunk 1]
```diff
@@ -0,0 +1,211 @@
...
+
+Result:
+
+```html
+<span>
+    <input type="text">
+</span>
+```
+
+When no specified, `div` tag is used.
```
Comment (samdark) [inline docs/guide/en/field-common-methods/base-field/container.md:23]:
```suggestion
When not specified, `div` tag is used.
```

File: docs/guide/en/field-common-methods/base-field/container.md [hunk 2]
```diff
@@ -0,0 +1,211 @@
...
+
+Result:
+
+```html
+<div class="field-container" data-type="name">
+    <input type="text">
+</div>
+```
+
+Note that values within the same attribute will not be merged, newly added overrides previous ones:
```
Comment (samdark) [inline docs/guide/en/field-common-methods/base-field/container.md:70]:
```suggestion
Note that values within the same attribute will not be merged, newly added value overrides previous ones:
```

File: docs/guide/en/field-common-methods/base-field/container.md [hunk 3]
```diff
@@ -0,0 +1,211 @@
...
+
+```html
+<div class="field-container">
+    <input type="text">
+</div>
+```
+
+No class is used by default.
+
+To add classes to the existing ones:
```
Comment (samdark) [inline docs/guide/en/field-common-methods/base-field/container.md:133]:
```suggestion
To add classes to existing ones:
```

File: docs/guide/en/field-common-methods/input-field/field.md [hunk 4]
```diff
@@ -0,0 +1,214 @@
+# Input field
+
+Input field is a parts field with exactly 1 input element.
```
Comment (samdark) [inline docs/guide/en/field-common-methods/input-field/field.md:3]: What is "parts field"?

File: docs/guide/en/field-common-methods/input-field/field.md [hunk 5]
```diff
@@ -0,0 +1,214 @@
...
+
+Result:
+
+```html
+<div>
+    <input type="text" class="input">
+</div>
+```
+
+To add attributes to the existing ones:
```
Comment (samdark) [inline docs/guide/en/field-common-methods/input-field/field.md:25]:
```suggestion
To add attributes to existing ones:
```

File: docs/guide/en/field-common-methods/input-field/field.md [hunk 6]
```diff
@@ -0,0 +1,214 @@
...
+
+Result:
+
+```html
+<div>
+    <input type="text" class="input" data-type="name">
+</div>
+```
+
+Note that values within the same attribute will not be merged, newly added overrides previous ones:
```
Comment (samdark) [inline docs/guide/en/field-common-methods/input-field/field.md:48]:
```suggestion
Note that values within the same attribute will not be merged, newly added value overrides previous ones:
```

File: docs/guide/en/field-common-methods/parts-field/field.md [hunk 7]
```diff
@@ -0,0 +1,336 @@
...
+```html
+<div>
+    <label class="label" data-type="label">Name</label>
+    <input type="text">
+    <div class="hint" data-type="hint">Enter name</div>
+    <div class="error" data-type="error">Name is not valid.</div>
+</div>
+```
+
+Note that values within the same attribute will not be merged, newly added overrides previous ones:
```
Comment (samdark) [inline docs/guide/en/field-common-methods/parts-field/field.md:167]:
```suggestion
Note that values within the same attribute will not be merged, newly added value overrides previous ones:
```

File: docs/guide/en/field-common-methods/parts-field/input-container.md [hunk 8]
```diff
@@ -0,0 +1,169 @@
...
+
+Result:
+
+```html
+<div>
+    <span><input type="text"></span>
+</div>
+```
+
+When no specified, no container is used for input.
```
Comment (samdark) [inline docs/guide/en/field-common-methods/parts-field/input-container.md:23]:
```suggestion
When not specified, no container is used for input.
```

File: docs/guide/en/field-common-methods/parts-field/input-container.md [hunk 9]
```diff
@@ -0,0 +1,169 @@
...
+
+## Attributes
+
+HTML attributes for outer container that wraps the input. Input container tag must be specified too in order for
```
Comment (samdark) [inline docs/guide/en/field-common-methods/parts-field/input-container.md:27]:
```suggestion
HTML attributes for outer container that wraps the input. Input container tag must be specified for
```

File: docs/guide/en/field-common-methods/parts-field/input-container.md [hunk 10]
```diff
@@ -0,0 +1,169 @@
...
+
+Result:
+
+```html
+<div>
+    <span class="input-container"><input type="text"></span>
+</div>
+```
+
+When no specified, no attributes are used for input.
```
Comment (samdark) [inline docs/guide/en/field-common-methods/parts-field/input-container.md:48]:
```suggestion
When not specified, no attributes are used for input.
```

File: docs/guide/en/field-common-methods/parts-field/input-container.md [hunk 11]
```diff
@@ -0,0 +1,169 @@
...
+
+Result:
+
+```html
+<div>
+    <span class="input-container" data-type="name"><input type="text"></span>
+</div>
+```
+
+Note that values within the same attribute will not be merged, newly added overrides previous ones:
```
Comment (samdark) [inline docs/guide/en/field-common-methods/parts-field/input-container.md:75]:
```suggestion
Note that values within the same attribute will not be merged, newly added value overrides previous ones:
```

File: docs/guide/en/field-common-methods/parts-field/input-container.md [hunk 12]
```diff
@@ -0,0 +1,169 @@
...
+
+```html
+<div>
+    <span class="focus"><input type="text"></span>
+</div>
+```
+
+## Class
+
+HTML class for outer container that wraps the input. Input container tag must be specified too in order for container to
```
Comment (samdark) [inline docs/guide/en/field-common-methods/parts-field/input-container.md:100]:
```suggestion
HTML class for outer container that wraps the input. Input container tag must be specified for container to
```

File: docs/guide/en/field-common-methods/parts-field/input-container.md [hunk 13]
```diff
@@ -0,0 +1,169 @@
...
+    ->inputContainerClass('input-container');
+```
+
+```html
+<div>
+    <span class="input-container"><input type="text"></span>
+</div>
+```
+
+When no specified, no class is used for input.
```
Comment (samdark) [inline docs/guide/en/field-common-methods/parts-field/input-container.md:119]:
```suggestion
When not specified, no class is used for input.
```

File: docs/guide/en/field-common-methods/parts-field/input-container.md [hunk 14]
```diff
@@ -0,0 +1,169 @@
...
+
+To add classes to the existing ones:
```
Comment (samdark) [inline docs/guide/en/field-common-methods/parts-field/input-container.md:121]:
```suggestion
To add classes to existing ones:
```

**Review by samdark — APPROVED** [review]

File: docs/guide/en/field-common-methods/input-field/field.md (hunk 4 above)
Comment (arogachev) [reply docs/guide/en/field-common-methods/input-field/field.md:3]: Maybe add link to parts field here?

**Review by vjik — CHANGES_REQUESTED** [review]
My suggestion: https://github.com/yiisoft/form/pull/337

File: docs/guide/en/field-common-methods/input-field/field.md (hunk 4 above)
Comment (samdark) [reply docs/guide/en/field-common-methods/input-field/field.md:3]: Yes :+1:

File: docs/guide/en/field-common-methods/parts-field/field.md [hunk 15]
```diff
@@ -0,0 +1,336 @@
+# Parts field
+
+Parts field consists of 4 elements:
+
+- Label;
+- Hint;
+- Input;
+- Error.
+
+Input here can be composite and can contain multiple elements.
```
Comment (samdark) [inline docs/guide/en/field-common-methods/parts-field/field.md:10]:
```suggestion
Input here can be composite and can contain multiple elements (parts).
```

File: docs/guide/en/field-methods.md [hunk 16]
```diff
@@ -0,0 +1,1618 @@
...
+```
+
+## `PartsField` based fields
+
+Parts field consists of 4 elements:
+
+- Label;
+- Hint;
+- Input;
+- Error.
```
Comment (vjik) [inline docs/guide/en/field-methods.md:176]:
```suggestion
- label;
- hint;
- input;
- error.
```

File: docs/guide/en/field-methods.md [hunk 17]
```diff
@@ -0,0 +1,1618 @@
...
+```
+
+```html
+<div>
+    <label>Name</label>
+    <input type="text">
+</div>
+```
+
+Another way is to use [config].
```
Comment (vjik) [inline docs/guide/en/field-methods.md:221]: Is missed link?

File: docs/guide/en/field-methods.md [hunk 18]
```diff
@@ -0,0 +1,1618 @@
...
+```
+
+```html
+<div>
+    <input type="text">
+    <div>Enter name</div>
+</div>
+```
+
+Another way is to use [config].
```
Comment (vjik) [inline docs/guide/en/field-methods.md:238]: Is missed link?

File: docs/guide/en/field-methods.md [hunk 19]
```diff
@@ -0,0 +1,1618 @@
...
+```
+
+```html
+<div>
+    <input type="text">
+    <div>Name is not valid.</div>
+</div>
+```
+
+Another way is to use [config].
```
Comment (vjik) [inline docs/guide/en/field-methods.md:255]: Is missed link?

File: docs/guide/en/field-methods.md [hunk 20]
```diff
@@ -0,0 +1,1618 @@
...
+}
+
+$field = $field->label('Name')
+
+echo $field;
+```
+
+### `labelConfig()`
+
+Config with [definitions](https://github.com/yiisoft/definitions) for label.
```
Comment (vjik) [inline docs/guide/en/field-methods.md:297]:
```suggestion
Config with [definitions](https://github.com/yiisoft/definitions) for `Label` widget.
```

File: docs/guide/en/field-methods.md [hunk 21]
```diff
@@ -0,0 +1,1618 @@
...
+```html
+<div>
+    <label class="label">Name</label>
+    <input type="text">
+</div>
+```
+
+### `hintConfig()`
+
+Config with [definitions](https://github.com/yiisoft/definitions) for hint.
```
Comment (vjik) [inline docs/guide/en/field-methods.md:314]:
```suggestion
Config with [definitions](https://github.com/yiisoft/definitions) for `Hint` widget.
```

File: docs/guide/en/field-methods.md [hunk 22]
```diff
@@ -0,0 +1,1618 @@
...
+```html
+<div>
+    <input type="text">
+    <div class="hint">Enter name</div>
+</div>
+```
+
+### `errorConfig()`
+
+Config with [definitions](https://github.com/yiisoft/definitions) for error.
```
Comment (vjik) [inline docs/guide/en/field-methods.md:331]:
```suggestion
Config with [definitions](https://github.com/yiisoft/definitions) for `Error` widget.
```

File: docs/guide/en/field-methods.md [hunk 23]
```diff
@@ -0,0 +1,1618 @@
...
+
+HTML attributes for label.
+
+```php
+echo \Yiisoft\Form\Field\Text::widget()
+    ->label('Name')
+    ->labelAttributes(['class' => 'label', 'data-type' => 'name']);
+```
+
+> The label must be visible (use [`label()`](#label) or [`labelConfig()](#labelconfig) methods).
```
Comment (vjik) [inline docs/guide/en/field-methods.md:356]:
```suggestion
> The label must be visible (use [`label()`](#label) or [`labelConfig()`](#labelconfig) methods).
```

File: docs/guide/en/field-methods.md [hunk 24]
```diff
@@ -0,0 +1,1618 @@
...
+
+HTML attributes for hint.
+
+```php
+echo \Yiisoft\Form\Field\Text::widget()
+    ->hint('Enter name')
+    ->hintAttributes(['class' => 'hint', 'data-type' => 'name']);
+```
+
+> The hint must be visible (use [`hint()`](#hint) or [`hintConfig()](#hintconfig) methods).
```
Comment (vjik) [inline docs/guide/en/field-methods.md:394]:
```suggestion
> The hint must be visible (use [`hint()`](#hint) or [`hintConfig()`](#hintconfig) methods).
```

File: docs/guide/en/field-methods.md [hunk 25]
```diff
@@ -0,0 +1,1618 @@
...
+
+HTML ID for label.
+
+```php
+echo \Yiisoft\Form\Field\Text::widget()
+    ->label('Name')
+    ->labelId('label');
+```
+
+> The label must be visible (use [`label()`](#label) or [`labelConfig()](#labelconfig) methods).
```
Comment (vjik) [inline docs/guide/en/field-methods.md:470]:
```suggestion
> The label must be visible (use [`label()`](#label) or [`labelConfig()`](#labelconfig) methods).
```

File: docs/guide/en/field-methods.md [hunk 26]
```diff
@@ -0,0 +1,1618 @@
...
+
+HTML ID for hint.
+
+```php
+echo \Yiisoft\Form\Field\Text::widget()
+    ->hint('Enter name')
+    ->hintId('hint');
+```
+
+> The hint must be visible (use [`hint()`](#hint) or [hintlConfig()](#hintconfig) methods).
```
Comment (vjik) [inline docs/guide/en/field-methods.md:489]:
```suggestion
> The hint must be visible (use [`hint()`](#hint) or [`hintConfig()`](#hintconfig) methods).
```

File: docs/guide/en/field-methods.md [hunk 27]
```diff
@@ -0,0 +1,1618 @@
...
+
+HTML ID for error.
+
+```php
+echo \Yiisoft\Form\Field\Text::widget()
+    ->error('Name is not valid.')
+    ->errorId('error');
+```
+
+> The error must be visible (use [`error()`](#error) or [errorlConfig()](#errorconfig) methods).
```
Comment (vjik) [inline docs/guide/en/field-methods.md:512]:
```suggestion
> The error must be visible (use [`error()`](#error) or [`errorConfig()`](#errorconfig) methods).
```

File: docs/guide/en/field-methods.md [hunk 28]
```diff
@@ -0,0 +1,1618 @@
...
+Result:
+
+```html
+<div>
+    <input type="text">
+    <div id="hint">Enter name</div>
+</div>
+```
+
+```labelClass```
```
Comment (vjik) [inline docs/guide/en/field-methods.md:500]:
```suggestion
```

File: docs/guide/en/field-methods.md [hunk 29]
```diff
@@ -0,0 +1,1618 @@
...
+
+HTML class for label. In case of multiple classes, pass them as separate arguments.
+
+```php
+echo \Yiisoft\Form\Field\Text::widget()
+    ->label('Name')
+    ->labelClass('label', 'focus');
+```
+
+> The label must be visible (use [`label()`](#label) or [`labelConfig()](#labelconfig) methods).:
```
Comment (vjik) [inline docs/guide/en/field-methods.md:533]:
```suggestion
> The label must be visible (use [`label()`](#label) or [`labelConfig()`](#labelconfig) methods).:
```

File: docs/guide/en/field-methods.md [hunk 30]
```diff
@@ -0,0 +1,1618 @@
...
+
+HTML class for hint. In case of multiple classes, pass them as separate arguments.
+
+```php
+echo \Yiisoft\Form\Field\Text::widget()
+    ->hint('Enter name')
+    ->hintClass('hint', 'focus');
+```
+
+> The hint must be visible (use [`hint()`](#hint) or [`hintConfig()](#hintconfig) methods).:
```
Comment (vjik) [inline docs/guide/en/field-methods.md:569]:
```suggestion
> The hint must be visible (use [`hint()`](#hint) or [`hintConfig()`](#hintconfig) methods).:
```

File: docs/guide/en/field-methods.md [hunk 31]
```diff
@@ -0,0 +1,1618 @@
...
+
+HTML class for error. In case of multiple classes, pass them as separate arguments.
+
+```php
+echo \Yiisoft\Form\Field\Text::widget()
+    ->error('Name is not valid.')
+    ->errorClass('error', 'focus');
+```
+
+> The error must be visible (use [`error()`](#error) or [`errorConfig()](#errorconfig) methods).:
```
Comment (vjik) [inline docs/guide/en/field-methods.md:605]:
```suggestion
> The error must be visible (use [`error()`](#error) or [`errorConfig()`](#errorconfig) methods).:
```

File: docs/guide/en/field-methods.md [hunk 32]
```diff
@@ -0,0 +1,1618 @@
...
+```html
+<div>
+    <span class="input-container focus primary"><input type="text"></span>
+</div>
+```
+
+### `template()`
+
+A template for a field where tokens (placeholders) are field parts. This template is used when field is created using
+`widget()` method.
```
Comment (vjik) [inline docs/guide/en/field-methods.md:722]:
```suggestion
A template for a field where tokens (placeholders) are field parts. This template is used when field is rendered.
```

File: docs/guide/en/field-methods.md [hunk 33]
```diff
@@ -0,0 +1,1618 @@
...
+
+Supported tokens:
```
Comment (vjik) [inline docs/guide/en/field-methods.md:724]:
```suggestion
Built-in tokens:
```

**Review by vjik — CHANGES_REQUESTED** [review]

**Review by vjik — APPROVED** [review]

**Pull request merged**